packages/keyring-controller/src/KeyringController.ts

```diff
@@ -723,39 +732,49 @@ export class KeyringController extends BaseController<
     }
   }
 
-  async unlockQRHardwareWalletAccount(index: number): Promise<void> {
+  async unlockQRHardwareWalletAccount(index: number): Promise<string> {
     const keyring = await this.getOrAddQRKeyring();
 
     keyring.setAccountToUnlock(index);
     const oldAccounts = await this.#keyring.getAccounts();
     await this.#keyring.addNewAccount(keyring);
     const newAccounts = await this.#keyring.getAccounts();
     this.updateIdentities(newAccounts);
+    let newHardwareWalletAddress = '';
     newAccounts.forEach((address: string) => {
       if (!oldAccounts.includes(address)) {
         if (this.setAccountLabel) {
           this.setAccountLabel(address, `${keyring.getName()} ${index}`);
         }
-        this.setSelectedAddress(address);
+        newHardwareWalletAddress = address;
       }
     });
     await this.#keyring.persistAllKeyrings();
     await this.fullUpdate();
+    return newHardwareWalletAddress;
   }
 
   async getAccountKeyringType(account: string): Promise<KeyringTypes> {
     return (await this.#keyring.getKeyringForAccount(account)).type;
   }
 
-  async forgetQRDevice(): Promise<void> {
+  async forgetQRDevice(): Promise<{
+    removedAccounts: string[];
+    remainingAccounts: string[];
+  }> {
     const keyring = await this.getOrAddQRKeyring();
+    const accountsIncludingHardware =
+      (await this.#keyring.getAccounts()) as string[];
     keyring.forgetDevice();
-    const accounts = (await this.#keyring.getAccounts()) as string[];
-    accounts.forEach((account) => {
-      this.setSelectedAddress(account);
-    });
+    const remainingAccounts = (await this.#keyring.getAccounts()) as string[];
+    const removedAccounts = accountsIncludingHardware.filter(
+      (address) => !remainingAccounts.includes(address),
+    );
+    this.updateIdentities(remainingAccounts);
     await this.#keyring.persistAllKeyrings();
     await this.fullUpdate();
+    // Return list of removed accounts as well as remaining accounts for the app to use.
```

Comment:
This comment doesn't seem to add any more information than the variables themselves; thoughts on removing?

Reply:
Agreed. Removed on [40dc630](https://github.com/MetaMask/core/pull/1095/commits/40dc630149c459e309a1f5368b5ee707cf59404f)